### PR TITLE
Alert History Screen

### PR DESCRIPTION
**Pull request by rinslow**
Screenshot from 2020-04-01 13-08-22
Screenshot from 2020-04-01 13-08-25

Solves https://github.com/Reznic/Inhalator/issues/18

### CODE REVIEW

File: data/alerts.py [hunk 1]
```diff
@@ -12,48 +17,103 @@ class AlertCodes(object):
...
-    def __init__(self, alert_code, additional_info=None):
+    error_dict = {
+        AlertCodes.PRESSURE_LOW: "Low Pressure",
+        AlertCodes.PRESSURE_HIGH: "High Pressure",
+        AlertCodes.VOLUME_LOW: "Low Volume",
+        AlertCodes.VOLUME_HIGH: "High Volume",
+        AlertCodes.NO_BREATH: "No Breathing",
+        AlertCodes.PEEP_TOO_HIGH: "High PEEP",
+        AlertCodes.PEEP_TOO_LOW: "Low PEEP",
+        AlertCodes.NO_CONFIGURATION_FILE: "Cannot Read Configuration"
```
Comment (osherdp) [inline data/alerts.py]: Maybe Configuration Not Found?

File: data/alerts.py [hunk 2]
```diff
@@ -12,48 +17,103 @@ class AlertCodes(object):
     NO_BREATH = 1 << 6
     NO_CONFIGURATION_FILE = 1 << 7
 
+    @classmethod
+    def is_valid(cls, alert_code):
+        return alert_code in map(int, list(cls))
 
 class Alert(object):
-    def __init__(self, alert_code, additional_info=None):
+    error_dict = {
```
Comment (osherdp) [inline data/alerts.py]: No need for the "dict"
`ALERT_CODE_TO_MESSAGE`

File: data/alerts.py [hunk 3]
```diff
@@ -12,48 +17,103 @@ class AlertCodes(object):
     NO_BREATH = 1 << 6
     NO_CONFIGURATION_FILE = 1 << 7
 
+    @classmethod
+    def is_valid(cls, alert_code):
+        return alert_code in map(int, list(cls))
```
Comment (osherdp) [inline data/alerts.py]: remove the `list`

File: data/alerts.py [hunk 4]
```diff
@@ -12,48 +17,103 @@ class AlertCodes(object):
...
+            self.timestamp = time.time()
+
+        else:
+            self.timestamp = timestamp
 
     def __eq__(self, other):
         return self.code == other
 
+    def __str__(self):
+        return self.error_dict.get(self.code, "Multiple Alerts")  # TODO: Elaborate
```
Comment (osherdp) [inline data/alerts.py]: what's the `self.code` value if there are multiple alerts

File: data/alerts.py [hunk 5]
```diff
@@ -12,48 +17,103 @@ class AlertCodes(object):
...
+        self.alerts_history = deque(maxlen=self.MAXIMUM_HISTORY_COUNT)
         self.queue = Queue(maxsize=self.MAXIMUM_ALERTS_AMOUNT)
         self.last_alert = Alert(AlertCodes.OK)
         self._subscribers = {}
 
     def __len__(self):
         return self.queue.qsize()
 
+    def history(self):
+        return list(self.alerts_history)
```
Comment (osherdp) [inline data/alerts.py]: will never be too large, right?
consider using queue with maxlen

**Comment by avrahamshukron** [comment]
Can you add screenshots to the PR description?

File: data/alerts.py (hunk 1 above)
Comment (rinslow) [reply data/alerts.py:33]: I think Configuration Error sums it up best

File: data/alerts.py (hunk 4 above)
Comment (rinslow) [reply data/alerts.py]: Solved for multiple alerts

File: data/alerts.py (hunk 5 above)
Comment (rinslow) [reply data/alerts.py:80]: I am using a queue with maxlen :rofl:

Comment (rinslow) [reply data/alerts.py:80]: See line 61

**Comment by rinslow** [comment]
> Can you add screenshots to the PR description?

sure

File: data/alerts.py [hunk 6]
```diff
@@ -12,48 +17,113 @@ class AlertCodes(object):
...
+        AlertCodes.NO_CONFIGURATION_FILE: "Configuration Error"
+    }
+
+    def __init__(self, alert_code, timestamp=None):
         self.code = alert_code
-        self.value = additional_info
+
+        if timestamp is None:
+            self.timestamp = time.time()
+
```
Comment (avrahamshukron) [inline data/alerts.py:41]:
```suggestion
            self.timestamp = time.time()
```

**Pull request merged**